tools/clutils/sd/saga-sd.1
.\" Man page for SAGA Service Discovery CLI
.TH SAGA-SD "1" "November 2009" "SAGA-SD " "User Commands"
.SH NAME
saga-sd \- performs Service Discovery queries via SAGA
.SH SYNOPSIS
.B saga-sd
[\fIOPTIONS\fR]
.SH DESCRIPTION
.\" Add any additional description here
.PP
Perform a Service Discovery query using three optional filters. Each filter should be enclosed in double quotes to avoid shell expansion.

Each of the filter strings uses a SQL92 syntax
as if it were part of a WHERE clause acting to select from a single table.
Multi-valued attributes are treated as a set of values.

Only the following operators are permitted in expressions:

IN, LIKE, AND, OR, NOT, =, >=, >, <=, <, <>

In addition: Column names,
             parentheses,
             column values as single quoted strings,
             numeric values
             and comma are permitted.

.PP
For a multi-valued attribute, the name of the attribute MUST be preceded
by the keyword ALL or ANY unless comparison with a set literal is intended.
For each part of the expression,
the attribute name MUST precede the literal value.

Results will be returned in a random order.

.SH OPTIONS
.TP
\fB\-\-context\fR \fICONTEXT\fR
specify the context. The keyord may be specified more than once to allow multiple contexts. 
Attributes may be specified for the context by adding a parenthesised, comma separated list of 
attribute value pairs, where the attribute and value are separated by an equals sign. Leading 
and trailing spaces are ignored. 
.TP
\fB\-\-sf\fR \fIFILTER\fR
only services matching this service FILTER will be returned. The default is no filtering on service attributes.
.TP
\fB\-\-df\fR \fIFILTER\fR
only services with a data section matching this data FILTER will be returned. The default is no filtering on service data.
.TP
\fB\-\-af\fR \fIFILTER\fR
only services matching this authorization FILTER will be returned. The default is derived from the user's SAGA context.
.TP
\fB\-n\fR \fINUM\fB \-\-count\fR \fINUM\fR
the NUM of services to display. The default is unlimited.
.TP
\fB\-\-sa\fR
display service attributes
.TP
\fB\-\-da\fR
display data attributes
.TP
\fB\-\-url\fR \fIURL\fR
the URL of the information system to use. This overrides any environment variables.
.TP
\fB\-h, --help\fR
display this help and exit
.SH Examples
.TP
saga-sd --context glite --sf "Site='RAL'" -n 1
Return the URL of the first service matching from the site RAL, where you have authorization to access the service using the glite context.
.TP
saga-sd --sf "Site IN ('INFN-CNAF', 'RAL-LCG2')" --af "" --sa
Return details of all services at INFN-CNAF and RAL-LCG2



CHANGES
                            SAGA C++ Core Libraries  
                    - The Simple API for Grid Applications -
                          http://www.saga-project.org


  Changes with SAGA Core Version 1.7                                    T.B.D
  ---------------------------------------------------------------------------
  o 

  Changes with SAGA C++ Version 1.6                         1. September 2011
  ---------------------------------------------------------------------------
  o Updated logging system
  o Implemented packaging mechanisms 
  o Complete list of tickets: http://bit.ly/nBMtep


  Changes with SAGA C++ Version 1.5.4                          19. April 2011
  ---------------------------------------------------------------------------
  o Fixes compatibility issue with boost::filesystem V3
    see: http://bit.ly/dFfrYO
  o http://faust.cct.lsu.edu/trac/saga/ticket/706
  o Some other minor fixes
  
    
  Changes with SAGA C++ Version 1.5.3                        10. January 2011
  ---------------------------------------------------------------------------
  o Lots of improvements in the documentation of the programming 
    examples and the programming example code itself
  o Addressed some Boost compatibility issues
  o Fixed tickets: http://bit.ly/h9XhIw


  Changes with SAGA C++ Version 1.5.2                      28. September 2010
  ---------------------------------------------------------------------------
  o Added support for Boost >= 1.43 to the build-system
  o Disabled the engine monitoring thread until the race cond. has been found
  o Fixed compilation problem with  external boost::asio
  o Fixed compilation problem with  external boost::filesystem and boost <= 1.35
  o Fixed problems that were cause by a bug in boost::filesystem 1.44
  o file.read() fixed in the default file adaptor
  o Lost of minor improvements and fixes


  Changes with SAGA C++ Version 1.5.1                         05. August 2010
  ---------------------------------------------------------------------------
  o Bugfix Release - fixes build system issues (include path precedence)
  o A detailed list of all closed tickets for this release can be 
    found here: http://tinyurl.com/2g2e8lb


  Changes with SAGA C++ Version 1.5                             09. July 2010
  ---------------------------------------------------------------------------
  o A detailed list of all closed tickets for this release can be 
    found here: http://tinyurl.com/2fa6euv
    

  Changes with SAGA C++ Version 1.4.2                          16. April 2010
  ---------------------------------------------------------------------------
  o We have completely restructured our code repositories. We have split up 
    SAGA into a "Core" (Engine, Packages, Default Adaptors) and into the 
    middleware adaptors. SAGA Core still lives in the same code repository,
    but the adaptors have been moved to:
    
    https://svn.cct.lsu.edu/repos/saga-adaptors.
    
    Beginning with this release, we will release Core, Python Bindings and 
    Adaptors seperately. This allows us to have individual release cycles
    for individual adaptors and components. One of the major proeblems we 
    had in the past was, that e.g. a single bug in the SSH adaptor lead to
    a completely new saga release. This won't happen anymore: now you will
    be able to download just the new release of that SSH adaptor and add it
    to your existing SAGA installation. 
    
    What do all these changes and restructuring mean for you? For you it means
    that you can't install SAGA from one source archive anymore. Yes, it 
    might have been convenient but it also caused a lot of confusions. If you
    want to install saga, you have to download and install "SAGA Core" first.
    After that, you can pick the adaptors you want and maybe the Python 
    bindings, download and install them. All Core, Adaptor and Bindings 
    releases can be found on the SAGA C++ download page:
    
    http://saga.cct.lsu.edu/software/cpp/download
    
  o Besides restructuring our code, we also found enough time to fix some
    of the bugs that were reported. Here's a list of recently closed tickets:
    
    http://tinyurl.com/y63aole
  

  Changes with SAGA C++ Version 1.4.1                       17. February 2010
  ---------------------------------------------------------------------------
   o A list of fixed bugs and added features can be found here:
     http://tinyurl.com/ycozhef

  Changes with SAGA C++ Version 1.4                         01. December 2009
  ---------------------------------------------------------------------------
   o A list of fixed bugs and added features can be found here:
     http://tinyurl.com/yk3f8yc

  Changes with SAGA C++ Version 1.3.3                        01. October 2009
  ---------------------------------------------------------------------------
   o EMERGENCY BUGFIX RELEASE
     Fixes broken context handling 
     Some small improvements in the ssh job adaptor

  Changes with SAGA C++ Version 1.3.2                      28. September 2009
  ---------------------------------------------------------------------------
   o We have improved our project structures. In order to create higher-
     quality releases, we agreed upon a set of strict release policies and
     a new, more conservative release branch. More information can be 
     found here: http://faust.cct.lsu.edu/trac/saga/wiki/Development
     This is the FIRST release coming out of this new branch! 

   o A list of fixed bugs and added features can be found here:
     http://tinyurl.com/ye2qusr

  Changes with SAGA C++ Version 1.3.1                         24. August 2009
  ---------------------------------------------------------------------------
  o A list of fixed bugs and added features can be found here:
    http://tinyurl.com/y8p4jng

  Changes with SAGA C++ Version 1.3                             22. June 2009

tools/clutils/sd/saga-sd.cpp
//  Distributed under the Boost Software License, Version 1.0. (See accompanying 
//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

#include <iostream>
#include <vector>
#include <string>
#include <limits>
#include <stdlib.h>

#include <boost/regex.hpp>

#include "saga/saga.hpp"
#include "saga/saga/sd.hpp"

bool compare_nocase(const std::string& first, const std::string& second) {
    size_t len = 0;
    while ((len < first.length()) && (len < second.length())) {
        if (std::tolower(first[len]) < std::tolower(second[len])) {
            return true;
        }

        else if (std::tolower(first[len]) > tolower(second[len])) {
            return false;
        }
        ++len;
    }

    if (first.length() < second.length()) {
        return true;
    }

    else {
        return false;
    }
}

/**
 * Function to return value from a string. A check is made that there is nothing else after valid input apart from white space.
 */
template<class T>
bool from_string(T& t, const std::string& s) {
    std::istringstream iss(s);
    if ((iss >> t).fail()) {
        return false;
    }
    int eof;
    iss >> eof;
    return iss.fail() && iss.eof();
}

/**
 * Get next parameter safely
 */
std::string getNext(int & optind, int argc, char *argv[]) {
    if (optind >= argc-1) {
        std::cerr << argv[optind] << " must be followed by a value" << std::endl;
        exit(1);
    }
    ++optind;
    return argv[optind];
}

/**
 * Trim blanks from each end of a string
 */
std::string trim(std::string& s) {   
    std::string::size_type os = s.find_first_not_of(" ");
    std::string::size_type cs = s.find_last_not_of(" ");
    return s.substr(os, cs-os+1);
}

/**
 * Main program
 */
int main(int argc, char *argv[]) {
#if !defined(SAGA_HAVE_PACKAGE_SD)
  std::cerr << "saga-sd: saga has been compiled without the sd "
               "package, bailing out" << std::endl;
  return -1;
#else
    std::string svcFilter = "";
    std::string authzFilter = "";
    std::string dataFilter = "";
    std::string serverUrl = ""; 
    std::vector<std::string> contexts;
    bool haveAuthzFilter = false;
    bool sa = false;
    bool da = false;
    int maxCount = (std::numeric_limits<int>::max)();

    int optind = 1;
    while ((optind < argc) && (argv[optind][0] == '-')) {
        std::string sw = argv[optind];

        if (sw == "--context") {
          contexts.push_back(getNext(optind, argc, argv));
        } else if (sw == "--sf") {
            svcFilter = getNext(optind, argc, argv);
        } else if (sw == "--af") {
            authzFilter = getNext(optind, argc, argv);
            haveAuthzFilter = true;
        } else if (sw == "--df") {
            dataFilter = getNext(optind, argc, argv);
        } else if (sw == "--url") {
            serverUrl = getNext(optind, argc, argv);
        } else if (sw == "--da") {
            da = true;
        } else if (sw == "--sa") {
            sa = true;
        } else if (sw == "-n" || sw == "--count") {
            std::string val = getNext(optind, argc, argv);
            if (!from_string<int> (maxCount, val) || maxCount <= 0) {
                std::cerr << "Count must be a positive integer and not: " << argv[optind] << std::endl;
                exit(1);
            }
        } else if (sw == "-h" || sw == "--help") {
            std::cout << std::endl;
            std::cout << "Usage: " << argv[0] << " [options]" << std::endl;
            std::cout << "Options:" << std::endl;
            std::cout << "   --context <context>   context (may be many) with optional attributes in form context1(p1=v1,p2=v2)" << std::endl;
            std::cout << "   --sf <service filter> the default is no filtering on service attributes" << std::endl;
            std::cout << "   --df <data filter>    the default is no filtering on service data" << std::endl;
            std::cout << "   --af <authz filter>   the default is derived from the context" << std::endl;
            std::cout << "   --count, -n           the number of services to display - default unlimited" << std::endl;
            std::cout << "   --sa                  display service attributes " << std::endl;
            std::cout << "   --da                  display data attributes" << std::endl;
            std::cout << "   --url <server url>    overrides any environment variables" << std::endl;
            std::cout << "   --help, -h            print this message" << std::endl;
            std::cout << std::endl;
            return 0;
        } else {
            std::cerr << "Unknown switch: " << argv[optind] << std::endl;
            std::cerr << "Run with -h for help" << std::endl;
            exit(1);
        }
        optind++;
    }

    saga::session s;
    std::vector<std::string>::const_iterator pos;
    for (pos = contexts.begin(); pos != contexts.end(); pos++) {
        std::string::size_type op,cp,ep;
        op = (*pos).find("(");
        cp = (*pos).rfind(")");
        if (op != std::string::npos && cp != std::string::npos) {
            std::string parms = (*pos).substr(op+1, cp-op-1);
            std::string contextName = (*pos).substr(0,op);
            contextName = trim(contextName);
            saga::context ctx(contextName);
            std::string::size_type here = 0;
            while ( (ep = parms.find("=", here))  != std::string::npos ) {
                std::string att = parms.substr(here, ep-here);
                att = trim(att);
                here = ep + 1;
                cp = parms.find(",", here);
                if (cp !=  std::string::npos) {
                    std::string value = parms.substr(here, cp-here);
                    value = trim(value);
                    here = cp + 1;
                    ctx.set_attribute(att,value);
                } else {
                    std::string value = parms.substr(here);
                    value = trim(value);
                    ctx.set_attribute(att,value);
                    break;
                }
            } ;
            s.add_context(ctx);
        } else {
            std::string contextName = *pos;
            contextName = trim(contextName);
            saga::context ctx(contextName);
            s.add_context(ctx);
        }
    }

    saga::sd::discoverer sd(s, serverUrl);
    std::vector<saga::sd::service_description> services;

    try {
        if (haveAuthzFilter) {
            services = sd.list_services(svcFilter, dataFilter, authzFilter);
        } else {
            services = sd.list_services(svcFilter, dataFilter);
        }
    }

    catch (saga::exception& e) {
        std::cerr << e.get_message() << std::endl << "Exiting." << std::endl;
        exit(1);
    }

    catch (std::exception& e) {
        std::cerr << "Caught std::exception: " << e.what() << std::endl;
        exit(1);
    }

    catch (...) {
        std::cerr << "Caught something unexpected." << std::endl;
        exit(1);
    }

    std::vector<saga::sd::service_description>::iterator servicesIter;
    std::vector<saga::sd::service_description>::const_iterator servicesEnd = services.end();

    for (servicesIter = services.begin(); servicesIter != servicesEnd && maxCount > 0; ++servicesIter, --maxCount) {
        using namespace saga::sd::attributes;

        if (da || sa) {
            std::cout << std::endl;
        }

        std::vector<std::string> attribNames = servicesIter->list_attributes();

        // Deal with the URL if it exists - it should!
        std::vector<std::string>::iterator pos = find(attribNames.begin(), attribNames.end(), "Url");
        if (pos == attribNames.end()) {
            std::cout << "Url : is not known" << std::endl;
        } else {
            std::cout << "Url : " << servicesIter->get_attribute("Url") << std::endl;
            attribNames.erase(pos);
        }

        if (sa) {
            //Sort our array of attribute names so they come out in a decent order
            sort(attribNames.begin(), attribNames.end(), compare_nocase);

            for (std::vector<std::string>::const_iterator attribNamesIter = attribNames.begin(); attribNamesIter
                    != attribNames.end(); ++attribNamesIter) {
                if (!servicesIter->attribute_is_vector(*attribNamesIter)) {
                    std::string attribValue = servicesIter->get_attribute(*attribNamesIter);
                    std::cout << "Service " << *attribNamesIter << " : " << attribValue << std::endl;
                }

                else {
                    std::vector<std::string> attribValues = servicesIter->get_vector_attribute(*attribNamesIter);
                    std::vector<std::string>::const_iterator attribValuesIter;
                    std::vector<std::string>::const_iterator attribValuesEnd = attribValues.end();
                    for (attribValuesIter = attribValues.begin(); attribValuesIter != attribValuesEnd; ++attribValuesIter) {
                        std::cout << "Service " << *attribNamesIter << " : " << *attribValuesIter << std::endl;
                    }
                }
            }
        }

        if (da) {
            saga::sd::service_data data;
            try {
                data = servicesIter->get_data();
            }

            //Not a lot we can do here so just skip it
            catch (saga::exception& e) {
                std::cerr << "EXCEPTION: " << e.get_message() << std::endl;
            }

            catch (std::exception& e) {
                std::cerr << "Caught std::exception: " << e.what() << std::endl;
            }

            catch (...) {
                std::cerr << "Caught something unexpected." << std::endl;
            }

            std::vector<std::string> dataNames = data.list_attributes();

            //Sort our array of data attribute names so they come out in a decent order
            sort(dataNames.begin(), dataNames.end(), compare_nocase);

            std::vector<std::string>::const_iterator dataIter;
            std::vector<std::string>::const_iterator dataEnd = dataNames.end();

            for (dataIter = dataNames.begin(); dataIter != dataEnd; ++dataIter) {
                if (!data.attribute_is_vector(*dataIter)) {
                    std::string dataValue;
                    dataValue = data.get_attribute(*dataIter);
                    std::cout << "Data " << *dataIter << " : " << dataValue << std::endl;
                }

                else {
                    std::vector<std::string> dataValues = data.get_vector_attribute(*dataIter);
                    std::vector<std::string>::const_iterator dataValuesIter;
                    std::vector<std::string>::const_iterator dataValuesEnd = dataValues.end();
                    for (dataValuesIter = dataValues.begin(); dataValuesIter != dataValuesEnd; ++dataValuesIter) {
                        std::cout << "Data " << *dataIter << " : " << *dataValuesIter << std::endl;
                    }
                }
            }
        }
    }
    if (da || sa) {
        std::cout << std::endl;
    }
    std::cout << services.size() << " services matched the query" << std::endl;
    if (da || sa) {
        std::cout << std::endl;
    }

    return 0;
#endif
}
